api/src/services/security-service.ts

```diff
@@ -76,4 +84,86 @@ export class SecurityService extends DBService {
       )
     );
   }
+
+  /**
+   * Returns the signed URL of a document for which a user has permissions.
+   *
+   * Rules:
+   * non-admin user cannot access the document when:
+   ** - it is pending review, OR
+   ** - user hasn't been granted an exception to every security rule
+   *
+   * non-admin user can access the document when:
+   * - document is not secured
+   * - user has the correct exceptions
+   *
+   *
+   * @param {number} artifactId
+   * @return {*}  {Promise<any>}
+   * @memberof SecurityService
+   */
+  async getSecuredArtifactBasedOnRulesAndPermissions(artifactId: number): Promise<any> {
+    const isSystemUserAdmin = await this.userService.isSystemUserAdmin();
+
+    const userId = this.connection.systemUserId();
+
+    const isDocumentPendingReview = (await this.artifactService.getArtifactById(artifactId)).security_review_timestamp
+      ? false
+      : true;
+
+    if (!isSystemUserAdmin && isDocumentPendingReview) {
+      throw new HTTP403('Documents that are pending review can only be downloaded by administrators.');
+    }
+
+    const documentSecurityRules = await this.getDocumentPersecutionAndHarmRules(artifactId);
+
+    const pers_harm_exceptions = await this.getPersecutionAndHarmExceptionsByUser(userId);
+
+    const userHasExceptionsToAllRules = documentSecurityRules.every((rule) => pers_harm_exceptions.includes(rule));
+
+    if (
+      !isSystemUserAdmin &&
+      !isDocumentPendingReview &&
+      documentSecurityRules.length > 0 &&
+      !userHasExceptionsToAllRules
+    ) {
+      throw new HTTP403('You do not have access to this document');
+    }
+
+    // access is granted because
+    // 1) admin
+    // 2) document is unsecured (not pending review, and has no security rules)
+    // 3) non-admin has exceptions all security rules
+    const response = await this.artifactService.getArtifactById(artifactId);
+    return getS3SignedURL(response.key);
+  }
+
+  /**
+   * Get the persecution or harm rules for which a user is granted exception
+   *
+   * @param {number} userId
+   * @return {*}  {Promise<number[]>}
+   * @memberof SecurityService
+   */
+  async getPersecutionAndHarmExceptionsByUser(userId: number): Promise<number[]> {
+    defaultLog.debug({ label: 'getPersecutionAndHarmExceptionsByUser' });
+
+    return (await this.securityRepository.getPersecutionAndHarmRulesExceptionsByUserId(userId)).map(
+      (item) => item.persecution_or_harm_id
+    );
```

Comment:
Not really critical: but I would potentially consider removing the map, and instead including it in the calling function. That way this function could satisfy anyone that wants the list of rules, but not necessarily just the IDs.  Same for the function below.

Reply:
Even though the repo functions only return the IDs right now, technically its all backwards compatible, so if we exapanded the SQL to return more fields, this function could still return an array of objects of a defined type, and the parent could still map out just the IDs.

Reply:
I guess there is a bit of a question around what the service level is really providing at that point, which is maybe a fair counter argument... Hhmm

Reply:
After all that, lets leave these functions as they are. Maybe just add the word `Id` to the function name, to make it clear that the response is just the Ids of the rules.